grpc-client-server/README.md

```diff
@@ -0,0 +1,188 @@
+Then update your server to execute the probe: do a HTTP fetch of `endpoint` the specified number of times.
+Use the standard [`net/http` package](https://pkg.go.dev/net/http).
+
+Time each fetch using the `time` package:
+
+```
+  	start := time.Now()
+	resp, _:= http.Get(endpoint)
+    elapsed := time.Since(start)
+```
+
+Add up all the elapsed times, divide by the number of repetitions, and return the average to the client.
+The client should print out the average value received.
+You can do arithmetic operations like addition and division on `time.Duration` values in Go.
+However, under the hood `time.Duration` is an `int64`. To divide by an `int32` and return a `float` will need to do some [type conversions](https://go.dev/tour/basics/13).
+
+## Add a client timeout
+
+Maybe the site you are probing is very slow, or the number of repetitions is very high.
+Either way, you don't want your program to wait forever. 
+
+On the client side, add a [timeout](https://pkg.go.dev/context#WithTimeout) to stop waiting after 1 second].
+
+Run your client against some website - how many repetitions do you need to see your client timeout?
+
+## Extra Challenges: Errors, Continuous Probing, Serve Prometheus Metrics
+
+These sections are optional - do them for an extra challenge if time permits.
+
+### Part 1:
+How do we know if the HTTP fetch succeeded at the server? Add a check to make sure it did.
+How should we deal with errors, i.e. if the endpoint isn't found?
+Modify your code and proto format to handle these cases.
+
+### Part 2: Run probes on a schedule
+Instead of getting probe times via the gRPC client we might consider running the endpoint probes on a schedule. Change the
+prober_server to accept a frequency argument instead of the old `repetitions` parameter, 
+and continue running probes every N seconds forever. 
+
+You will need to use a [goroutine](https://gobyexample.com/goroutines) to keep running probes in the background.
+
+Now, add a new client operation to get the value of the last probe run.
+To make this work, you will need to pass a value to your prober goroutine that it can use to store the most recent latency.
+You can store this value in your `server` struct.
+You will also need to use a `sync.Mutex` to guarantee that the goroutine that serves your client request sees the last value 
+written by the prober goroutine. A Tour of Go has a good example of [using a sync.Mutex](https://go.dev/tour/concurrency/9).
+
+
+### Part 3: Add Prometheus Metrics
+If you still have time, then let's learn something about how to monitor applications.
+
+In software operations, we want to know what our software is doing and how it is performing.
+One very useful technique is to have your program export metrics. Metrics are basically values that your 
+program makes available (via HTTP). Specialised programs, such as Prometheus, can then fetch metrics regularly
```

Comment:
This draws a metrics-HTTP link, as if metrics _have to be_ exposed over HTTP. Is that correct, or a implementation detail here? Perhaps `(via HTTP)` → `to external systems such as...`